File: 알고리즘 기본 구현/EdmondsKarp.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.IO;

namespace EdmondsKarp {
    class Endpoint {
        public int Number { get; init; }
        public List<DirectedEdge> Outgoing { get; set; }

        public Endpoint(int num) {
            Number = num;
            Outgoing = new List<DirectedEdge>();
        }
    }

    class DirectedEdge {
        public static readonly DirectedEdge Dummy = new DirectedEdge();

        public Endpoint Source { get; init; }
        public Endpoint Destination { get; init; }
        public long Capacity { get; set; } = 0;
        public long Flow { get; private set; } = 0;
        public DirectedEdge Inverse { get; private set; }

        public long Spare => Capacity - Flow;

        public DirectedEdge(Endpoint src, Endpoint dst, long capacity) {
            Source = src;
            Destination = dst;
            Capacity = capacity;
            Inverse = new DirectedEdge(this);

            Source.Outgoing.Add(this);
        }

        public DirectedEdge(Vertex src, Vertex dst, long capacity)
            : this(src.Out, dst.In, capacity) {}

        private DirectedEdge(DirectedEdge inverse) {
            Source = inverse.Destination;
            Destination = inverse.Source;
            Inverse = inverse;

            Source.Outgoing.Add(this);
        }

        private DirectedEdge() {}

        public void AddFlow(long add) {
            Flow += add;
            Inverse.Flow -= add;
        }

        public void ClearFlow() {
            Flow = 0;
            Inverse.Flow = 0;
        }
    }

    class Vertex {
        public int Number { get; init; }
        public Endpoint In { get; init; }
        public Endpoint Out { get; init; }
        public DirectedEdge InnerEdge { get; init; }

        public Vertex(int num, long capacity = long.MaxValue/2) {
            Number = num;
            In = new Endpoint(num*2);
            Out = new Endpoint(num*2+1);
            InnerEdge = new DirectedEdge(In, Out, capacity);
        }
    }

    class Graph {
        private Vertex[] vertices;

        public Graph(int verCount) {
            vertices = new Vertex[verCount];
            for (int i = 0; i < verCount; ++i) {
                vertices[i] = new Vertex(i);
            }
        }

        public long GetInnerFlow(int verIdx) {
            return vertices[verIdx].InnerEdge.Flow;
        }

        public void SetInnerCapacity(int verIdx, long capacity) {
            vertices[verIdx].InnerEdge.Capacity = capacity;
        }

        public void AddEdge(int src, int dst, long capacity = long.MaxValue/2) {
            new DirectedEdge(vertices[src], vertices[dst], capacity);
        }

        private bool FindAugmentingPath(Vertex src, Vertex dst, ref DirectedEdge[] prev) {
            Array.Fill(prev, null);
            Queue<Endpoint> q = new Queue<Endpoint>();
            q.Enqueue(src.In);
            prev[src.In.Number] = DirectedEdge.Dummy;

            while (q.Count > 0) {
                var curr = q.Dequeue();

                foreach (var edge in curr.Outgoing) {
                    if (0 < edge.Spare && prev[edge.Destination.Number] == null) {
                        q.Enqueue(edge.Destination);
                        prev[edge.Destination.Number] = edge;
                    }
                    if (prev[dst.Out.Number] != null) {
                        return true;
                    }
                }
            }

            return false;
        }

        public long EdmondsKarp(int src, int dst) {
            ClearFlows();
            Vertex vs = vertices[src], vd = vertices[dst];
            DirectedEdge[] prev = new DirectedEdge[vertices.Length * 2];
            long result = 0;

            while (FindAugmentingPath(vs, vd, ref prev)) {
                long add = long.MaxValue;

                for (var edge = prev[vd.Out.Number]; edge != DirectedEdge.Dummy; edge = prev[edge.Source.Number]) {
                    add = Math.Min(add, edge.Spare);
                }
                for (var edge = prev[vd.Out.Number]; edge != DirectedEdge.Dummy; edge = prev[edge.Source.Number]) {
                    edge.AddFlow(add);
                }

                result += add;
            }

            return result;
        }

        public void ClearFlows() {
            foreach (var v in vertices) {
                foreach (var e in v.In.Outgoing) {
                    e.ClearFlow();
                }
                foreach (var e in v.Out.Outgoing) {
                    e.ClearFlow();
                }
            }
        }
    }
}

File: 알고리즘 기본 구현/CPP/EdmondsKarp.cpp
#include <algorithm>
#include <vector>
#include <queue>

const int INF = INT32_MAX/2;

struct Edge {
    int src, dst, flow, capacity;
    Edge* opposite;

    Edge(int src, int dst, int capacity)
        : src(src), dst(dst), flow(0), capacity(capacity) {}

    void addFlow(int add) {
        flow += add;
        if (opposite) {
            opposite->flow -= add;
        }
    }

    int getSpare() {
        return capacity-flow;
    }
};

class Graph {
private:
    std::vector<std::vector<Edge*>> graph;

    bool findAugPath(int src, int dst, std::vector<Edge*>& path) {
        std::vector<Edge*> prev(graph.size(), nullptr);
        std::queue<int> q;
        q.push(src);

        while (!q.empty() && prev[dst] == nullptr) {
            int u = q.front();
            q.pop();

            for (Edge* e : graph[u]) {
                int v = e->dst;
                if (e->getSpare() > 0 && v != src && u != v && prev[v] == nullptr) {
                    q.push(v);
                    prev[v] = e;
                    if (prev[dst]) {
                        break;
                    }
                }
            }
        }

        if (prev[dst]) {
            path.resize(0);
            for (Edge* e = prev[dst]; e != nullptr; e = prev[e->src]) {
                path.push_back(e);
            }
            return true;
        }
        return false;
    }

public:
    Graph(int vertexCount) {
        graph.resize(vertexCount);
    }

    ~Graph() {
        for (auto& edges : graph) {
            for (Edge* e : edges) {
                if (e) { delete e; }
            }
        }
    }

    void addEdge(int src, int dst, int capacity) {
        Edge* curr = new Edge(src, dst, capacity);
        Edge* oppo = new Edge(dst, src, 0);
        curr->opposite = oppo;
        oppo->opposite = curr;
        graph[src].push_back(curr);
        graph[dst].push_back(oppo);
    }

    int getMaxFlow(int src, int dst) {
        int result = 0;
        std::vector<Edge*> path;

        while (findAugPath(src, dst, path)) {
            int add = INT32_MAX;

            for (Edge* e : path) {
                add = std::min(add, e->getSpare());
            }
            for (Edge* e : path) {
                e->addFlow(add);
            }

            result += add;
        }

        return result;
    }
};